New: Write metadata to tags, with UI for previewing changes

ta264 opened the pull request:
#### Database Migration
NO

#### Description
Enables writing metadata to audio files once they have been matched.

Four levels are currently available:
- Off
- New downloads only
- All files
- All files, and keep in sync with updates from MusicBrainz

Existing files will not be edited unless requested in the artist/album/mass editor page.

Optionally you can scrub the other existing tags, removing anything not set by Lidarr.

#### Todos
- [x] Tests

#### Issues Fixed or Closed by this PR

* Fixes #289 
* Fixes #603 
* Fixes #639 
* Fixes #641

Qstick commented on frontend/src/Retag/RetagPreviewRow.js:
```diff
@@ -0,0 +1,103 @@
+
+function formatMissing(value) {
+  if (value === undefined || value === 0 || value === '0') {
+    return (<Icon name={icons.BAN} size="12" />);
```
`size={12}` Pass size as number instead of string so Icon prop type checks are happy

Qstick commented on frontend/src/Retag/RetagPreviewRow.js:
```diff
@@ -0,0 +1,103 @@
+
+function formatChange(oldValue, newValue) {
+  return (
+    <div> {formatMissing(oldValue)} <Icon name={icons.ARROW_RIGHT_NO_CIRCLE} size="12" /> {formatMissing(newValue)} </div>
```
`size={12}`

Qstick commented on frontend/src/Retag/RetagPreviewRow.js:
```diff
@@ -0,0 +1,103 @@
+import styles from './RetagPreviewRow.css';
+import DescriptionList from 'Components/DescriptionList/DescriptionList';
+import DescriptionListItem from 'Components/DescriptionList/DescriptionListItem';
+import DescriptionListItemTitle from 'Components/DescriptionList/DescriptionListItemTitle';
```
Not Used

Qstick commented on frontend/src/Retag/RetagPreviewRow.js:
```diff
@@ -0,0 +1,103 @@
+import DescriptionList from 'Components/DescriptionList/DescriptionList';
+import DescriptionListItem from 'Components/DescriptionList/DescriptionListItem';
+import DescriptionListItemTitle from 'Components/DescriptionList/DescriptionListItemTitle';
+import DescriptionListItemDescription from 'Components/DescriptionList/DescriptionListItemDescription';
```
Not used

Qstick commented on frontend/src/Retag/RetagPreviewRow.js:
```diff
@@ -0,0 +1,103 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import { icons, kinds } from 'Helpers/Props';
```
`kinds` is not used

Qstick commented on frontend/src/Retag/RetagPreviewModalContentConnector.js:
```diff
@@ -0,0 +1,98 @@
+      albumId,
+      retagTracks,
+      isPopulated,
+      isFetching
```
`retagTracks`, `isPopulated`, and `isFetching` should be in props validation

Qstick commented on frontend/src/Settings/MediaManagement/MediaManagement.js:
```diff
@@ -265,6 +272,35 @@ class MediaManagement extends Component {
+                          {...settings.scrubAudioTags}
+                        />
+                      </FormGroup>
+                  }
```
Does this make more sense in Metadata section? Thats where all metadata file options are (images, nfos, etc..,)

Qstick commented on src/NzbDrone.Core/MediaFiles/AudioTagService.cs:
```diff
@@ -0,0 +1,333 @@
+
+        private void UpdateTrackfileSize(TrackFile trackfile, string path)
+        {
+            // update the saved tracksize so that the importer doesn't get confused on the next scan
```
`file size` or `track size` instead of `tracksize`

ta264 commented:
I've moved the settings into the metadata section.  Let me know what you think - if you like it like that some of the stuff should get renamed from the specific `MetadataProvider*` to something more general

Qstick commented:
We should likely add a "More Info" wiki link below the setting like this:

![image](https://user-images.githubusercontent.com/376117/54001809-8c481480-411a-11e9-999f-5ff7716a5ba6.png)

Also, what do you feel about defaulting to new downloads only? Maybe less risky for the time being.

Qstick commented:
Also have +1 test failure compared to develop. Maybe a fluke test though.

ta264 commented:
Test failure fixed.  I think I've got it set up so that metadata writing is completely disabled by default because the I think default is to hardlink torrents and you don't want to be altering seeded files

Qstick commented on frontend/src/Settings/Metadata/MetadataProvider/MetadataProvider.js:
```diff
@@ -54,6 +61,37 @@ function MetadataProvider(props) {
+              {
+                settings.writeAudioTags.value !== 'no' &&
+                  <FormGroup>
+                    <FormLabel>Scrub Existing Metadata from Audio Files</FormLabel>
```
Can we shorten up this label to something like `Scrub Existing Tags`, currently it wraps in UI. Help text will explain more if user needs it.

![image](https://user-images.githubusercontent.com/376117/54065269-51161600-41ec-11e9-8022-ffee9dc2232d.png)

Qstick commented:
Should we allow manual re-tags even if setting is set to new downloads only or no tagging?

Qstick commented on frontend/src/Organize/OrganizePreviewModalContent.js:
```diff
@@ -107,13 +106,7 @@ class OrganizePreviewModalContent extends Component {
-                    <div>Renaming is disabled, nothing to rename</div>
-                }
-              </div>
+              <div>Success! My work is done, no files to rename.</div>
```
Don't know that we should do the same thing for renaming as for tagging. Renaming has a single checkbox that hides naming settings if turned off. Thus user is going to be renaming something that they have no control of the format. If we do this we should likely always show the track format naming field in settings whether renaming is enabled or not.

ta264 commented on src/NzbDrone.Core/MediaFiles/AudioTagService.cs:
```diff
@@ -0,0 +1,357 @@
+
+        public void WriteTags(TrackFile trackfile, bool newDownload)
+        {
+            if (_configService.WriteAudioTags == WriteAudioTagsType.No ||
```
Yeah, sorry, my change for this yesterday was half baked, will address properly

Qstick commented:
Do we need to do anything about the track number tag, does beets add the leading zero as Lidarr does above?

ta264 commented:
I've fixed the rogue leading zero in flac tags.  Also sorted the mapping to `ReleaseStatus` and `ReleaseType` and lowercased these as picard does.  Also corrected handling of dates in id3 tags.

Qstick commented:
@ta264 in regards to the ReleaseStatus and ReleaseType, do we want to check/write both options given beets writes it to the same field as Taglib default? I have a feeling alot of users will have both ways if that's the case. 

https://github.com/beetbox/beets/blob/e09e6976074f9167173c50e3443e3cf00aa5f48c/beets/mediafile.py#L1801
https://github.com/beetbox/beets/blob/e09e6976074f9167173c50e3443e3cf00aa5f48c/beets/mediafile.py#L1731

ta264 commented:
Yeah, that's probably a better idea. Should be easy enough

ta264 commented:
Hopefully this is good to go now.

Qstick commented:
Looks good. Looks like needs a quick fix on a merge conflict due to 36f34a41137b22d34a2cd9173beea7f9bbe90d24

ebnozal commented:
For some reason it does not look like it is including genres when it writes the metadata for the track

ta264 commented:
That's because we don't have genre data in lidarr. It's on the list.

Qstick commented:
Soon! When MB makes the genre list a db table, which they plan to release on May 13 as part of the Q2 database schema update
